spec/services/projects/create_artifact_work_package_service_spec.rb

```diff
@@ -157,5 +158,75 @@
       expect(artifact_work_package.last_journal.notes).to include(/<mention[^>]+>@#{assignee_user.name}<\/mention>/)
     end
+
+    context "when artifact storage is internal" do
+      it "attaches directly to the work package" do
+        project.update(project_creation_wizard_artifact_export_type: "attachment")
+        result = instance.call
+        project = result.result
+
+        artifact_work_package = WorkPackage.find(project.project_creation_wizard_artifact_work_package_id)
+        expect(artifact_work_package.attachments.count).to eq(1)
+        attachment = artifact_work_package.attachments.first
+        date = Time.zone.today.iso8601
+        expect(attachment.content_type).to eq "application/pdf"
+        expect(attachment.filename).to match /Important_Project_project_mandate_#{date}_\d+-\d+.pdf/
+      end
+    end
+
+    context "when artifact storage is project storage" do
+      let(:storage) { create(:nextcloud_storage_with_local_connection) }
+      let(:project_storage) { create(:project_storage, project:, storage:, project_folder_id: "/project_folder") }
+
+      let(:service_result) { ServiceResult.success(result: nil) }
+
+      before do
+        project.update(
+          project_creation_wizard_artifact_export_type: "file_link",
+          project_creation_wizard_artifact_export_storage: project_storage.id
+        )
+
+        allow(Storages::UploadFileService)
+          .to receive(:call)
+          .and_return(service_result)
+      end
+
+      it "calls the nextcloud storage service" do
+        result = instance.call
+        project = result.result
+
+        expect(result).to be_success
+        artifact_work_package = WorkPackage.find(project.project_creation_wizard_artifact_work_package_id)
+        expect(artifact_work_package.attachments.count).to eq(0)
+
+        date = Time.zone.today.iso8601
+        expect(Storages::UploadFileService)
+          .to have_received(:call)
+          .with(container: project,
+                project_storage:,
+                file_path: "project_mandate",
+                filename: /Important_Project_project_mandate_#{date}_\d+-\d+.pdf/,
+                file_data: instance_of(StringIO))
+      end
+
+      context "when service call fails" do
+        let(:service_result) { ServiceResult.failure(result: nil) }
+
+        it "rolls back the work package" do
+          result = instance.call
+          project = result.result
+
+          expect(Storages::UploadFileService)
+            .to have_received(:call)
+
+          # The outer service is successful, but an error is added
+          expect(result).to be_success
+          expect(result.errors[:base]).to include "Failed to store artifact in file storage"
+
+          artifact_work_package = WorkPackage.find(project.project_creation_wizard_artifact_work_package_id)
+          expect(artifact_work_package.attachments.count).to eq(0)
```

Comment:
Is it ok as a behavior? How would the artifact upload be triggered again?

Reply:
It's probably not okay, but I'm not sure how to handle it better.

it can be triggered manually from the project, and we'll need some way to trigger it again from the wizard. Addiitionally, it should be recreated whenever the WPs status changes